File: app/settings/Selector.lua
local ____exports = {}
local ____utils = require("utils.utils")
local get_nodes = ____utils.get_nodes
local ____Tween = require("utils.defoldTweens.Tween")
local tween = ____Tween.tween
function ____exports.Selector(druid, template_name, is_bg_off)
    if is_bg_off == nil then
        is_bg_off = false
    end
    local select, is_selected, anim
    function select(enabled)
        is_selected = enabled
        local anim_key = enabled and "select" or "unselect"
        for ____, tween in ipairs(anim[anim_key]) do
            tween.start(false)
        end
    end
    local node_clone = gui.clone_tree(gui.get_node(template_name .. "/btn"))
    local nodes = get_nodes({"bg", "sel", "icon", "btn"}, template_name, node_clone)
    local btns = druid:new_button(
        nodes.btn,
        function() return select(true) end
    )
    is_selected = false
    gui.set_enabled(nodes.btn, true)
    if is_bg_off then
        gui.set_enabled(nodes.bg, false)
    end
    anim = {
        select = {tween(nodes.sel).opacityTo(0, 0).opacityTo(0.1, 1)},
        unselect = {tween(nodes.sel).opacityTo(0.1, 0)}
    }
    return {
        on = function(callback) return btns.on_click:subscribe(callback) end,
        root_node = nodes.btn,
        nodes = nodes,
        select = select,
        druid_node = btns
    }
end
return ____exports

File: src/settings/Selector.ts
/* eslint-disable @typescript-eslint/no-unsafe-argument */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { get_nodes } from '../utils/utils';
import { tween } from '../utils/defoldTweens/Tween';

export function Selector(druid: DruidClass, template_name: string, is_bg_off = false) {
    const node_clone = gui.clone_tree(gui.get_node(template_name + '/btn'));
    const nodes = get_nodes(['bg', 'sel', 'icon', 'btn'] as const, template_name, node_clone);
    const btns = druid.new_button(nodes.btn, () => select(true));
    let is_selected = false;
    gui.set_enabled(nodes.btn, true);
    if (is_bg_off)
        gui.set_enabled(nodes.bg, false);

    const anim = {
        select: [
            tween(nodes.sel)
                .opacityTo(0, 0)
                .opacityTo(0.1, 1),
        ],
        unselect: [
            tween(nodes.sel)
                .opacityTo(0.1, 0),
        ]
    };

    function select(enabled: boolean) {
        is_selected = enabled;

        const anim_key: keyof typeof anim = enabled ? 'select' : 'unselect';
        for (const tween of anim[anim_key])
            tween.start(false);
    }

    return {
        on: (callback: () => void) => btns.on_click.subscribe(callback), root_node: nodes.btn, nodes, select, druid_node: btns
    };
}
